src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,394 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.Clustering;
+using Hazelcast.Core;
+using Hazelcast.DistributedObjects;
+using Hazelcast.Exceptions;
+using Hazelcast.Protocol;
+using Hazelcast.Protocol.Codecs;
+using Hazelcast.Protocol.Models;
+
+namespace Hazelcast.CP
+{
+    /// <summary>
+    /// Implementation of <see cref="IFencedLock"/>
+    /// </summary>
+    internal partial class FencedLock : CPDistributedObjectBase, IFencedLock
+    {
+        /// <summary>
+        /// Gets the current context identifier.
+        /// </summary>
+        /// <remarks>
+        /// Hazelcast APIs call this the thread identified and maintain locks "per threads",
+        /// so we are keeping the name here internally, but in reality this is not a thread
+        /// identifier anymore - it is attached to the async context so it can flow with
+        /// async operations.
+        /// </remarks>
+        private static long ContextId => AsyncContext.Current.Id;
+        private readonly ConcurrentDictionary<long, long> _lockedThreadToSession = new ConcurrentDictionary<long, long>();
+        private readonly CPSubsystemSession _subsystemSession;
+        private readonly CPGroupId _groupId;
+        private int _destroyed;
+
+        public const long InvalidFence = 0;
+        ICPGroupId ICPDistributedObject.GroupId => _groupId;
+        long IFencedLock.InvalidFence => InvalidFence;
+
+        public FencedLock(string name, CPGroupId groupId, Cluster cluster, CPSubsystemSession subsystemSession) : base(ServiceNames.FencedLock, name, groupId, cluster)
+        {
+            _subsystemSession = subsystemSession;
+            _groupId = groupId;
+            HConsole.Configure(x => x.Configure<FencedLock>().SetIndent(2).SetPrefix("FENCEDLOCK"));
+        }
+
+        #region IFencedLock Methods
+        /// <inheritdoc/>  
+        public async Task<long> GetFenceAsync()
+        {
+            var threadId = ContextId;
+            var sessionId = _subsystemSession.GetSessionId(CPGroupId);
+            HConsole.WriteLine(this, $"Session:{sessionId} ThreadId:{threadId} -> GetFence");
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            if (sessionId == CPSubsystemSession.NoSessionId)
+            {
+                _lockedThreadToSession.TryRemove(threadId, out var _);
+                throw new SynchronizationLockException();
```

Comment:
why different exception from Java? Also the exception message string.

Reply:
This is the `IllegalMonitorException `counterpart on .NET.